numpy/core/src/multiarray/arraytypes.c.src

```diff
@@ -4610,20 +4508,75 @@ set_typeinfo(PyObject *dict)
      *          CFLOAT, CDOUBLE, CLONGDOUBLE,
      *          OBJECT, STRING, UNICODE, VOID,
      *          DATETIME,TIMEDELTA#
+     * #NAME = Bool,
+     *         Byte, UByte, Short, UShort, Int, UInt,
+     *         Long, ULong, LongLong, ULongLong,
+     *         Half, Float, Double, LongDouble,
+     *         CFloat, CDouble, CLongDouble,
+     *         Object, String, Unicode, Void,
+     *         Datetime, Timedelta#
+     * #kind = GENBOOL,
+     *         SIGNED, UNSIGNED, SIGNED, UNSIGNED, SIGNED, UNSIGNED,
+     *         SIGNED, UNSIGNED, SIGNED, UNSIGNED,
+     *         FLOATING, FLOATING, FLOATING, FLOATING,
+     *         COMPLEX, COMPLEX, COMPLEX,
+     *         OBJECT, STRING, UNICODE, VOID,
+     *         DATETIME, TIMEDELTA#
+     * #endian = |*3, =*15, |, |, =, |, =*2#
+     * #flags= 0*18, NPY_OBJECT_DTYPE_FLAGS,0, NPY_NEEDS_INIT, 0*3#
+     * #fromtype = npy_bool,
+     *             npy_byte, npy_ubyte, npy_short, npy_ushort, npy_int, npy_uint,
+     *             npy_long, npy_ulong, npy_longlong, npy_ulonglong,
+     *             npy_half, npy_float, npy_double, npy_longdouble,
+     *             npy_cfloat, npy_cdouble, npy_clongdouble,
+     *             PyObject *, char, npy_ucs4, char,
+     *             npy_datetime, npy_timedelta#
+     * #DO_ELSIZE = 1*19, 0*3, 1*2#
      */
 
-    /**begin repeat1
-     *
-     * #name2 = HALF, DATETIME, TIMEDELTA#
-     */
+    dtype = PyObject_New(PyArray_Descr, &PyArrayDescr_Type);
+#ifndef USE_DTYPE_AS_PYOBJECT
+    /* Don't copy PyObject_HEAD part */
+    memset((char *)dtype + sizeof(PyObject),
```

Comment:
For me (and thus for future people looking at this...), it would be helpful to expand on why this is done. Does it get filled out later by actually instantiating? I realize links to documentation can get out of sync, but might still be better than nothing.

Looking at the documentation for `PyObject_New`, I see "Fields not defined by the Python object header are not initialized" - so, is the point here to undo the initialization that is done? But why?

Reply:
My question was whether it should be zeroed: presumably `PyObject_New` initializes this section (since it states that other parts are not initialized, suggesting that these parts are), and, again presumably, this is with stuff from `PyArrayDescr_Type`. If those presumptions hold, are those items wrong? Or might they have, e.g., init/new functions that can just be used?

Reply:
No, `PyObject_New` does not initialize this section, because it's not part of the `PyObject` head - in the same way that `PyObject_New` does not initialize new instance of `PyArrayObject`

Reply:
The comment here is wrong though - the word `copy` is no longer appropriate.

Reply:
Arggh, I just looked so poorly - of course the code zeros everything beyond `PyObject`. Which definitely makes sense.